### PR TITLE
Better groupBy error messages and docs around resource limits.

### PR DESCRIPTION
**Pull request by gianm**
Improvement on #4046, based on complaints that it was hard to understand how to tune groupBy v2 for a particular query load.

### CODE REVIEW

**Comment by fjy** [comment]
👍

File: processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java [hunk 1]
```diff
@@ -59,6 +59,16 @@
 public class BufferGrouper<KeyType> implements Grouper<KeyType>
 {
   private static final Logger log = new Logger(BufferGrouper.class);
+  private static final AggregateResult DICTIONARY_FULL = AggregateResult.failure(
+      "Not enough dictionary space to execute this query. Try increasing "
+      + "druid.query.groupBy.maxMergingDictionarySize or enable disk spilling by setting "
+      + "druid.query.groupBy.maxOnDiskStorage to a positive number."
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java]: s/positive/larger , user might already have a positive number in there

File: processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java [hunk 2]
```diff
@@ -59,6 +59,16 @@
...
+  );
+  private static final AggregateResult HASHTABLE_FULL = AggregateResult.failure(
+      "Not enough aggregation table space to execute this query. Try increasing "
+      + "druid.processing.buffer.sizeBytes or enable disk spilling by setting "
+      + "druid.query.groupBy.maxOnDiskStorage to a positive number."
```
Comment (himanshug) [inline processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java]: larger

**Comment by himanshug** [comment]
:+1: aside from minor doc comments

File: docs/content/querying/groupbyquery.md [hunk 3]
```diff
@@ -151,6 +154,34 @@ historical nodes.
...
+- druid.query.groupBy.maxMergingDictionarySize: size of the on-heap dictionary used when grouping on strings, per query,
+in bytes. Note that this is based on a rough estimate of the dictionary size, not the actual size.
+- druid.query.groupBy.maxOnDiskStorage: amount of space on disk used for aggregation, per query, in bytes. By default,
+this is 0, which means all aggregation is done in-memory.
+
+If maxOnDiskStorage is 0 (the default) then a query that exceeds either the on-heap dictionary limit, or the off-heap
+aggregation table limit, will fail with a "Resource limit exceeded" error describing the limit that was exceeded. If
+maxOnDiskStorage is greater than 0, queries that exceed the in-memory limits will start using disk for aggregation.
+Queries that then also exceed maxOnDiskStorage will fail with a "Resource limit exceeded" error indicating that they
+ran out of disk space.
```
Comment (gianm) [inline docs/content/querying/groupbyquery.md]: Clarified.

File: docs/content/querying/groupbyquery.md [hunk 4]
```diff
@@ -151,6 +154,34 @@ historical nodes.
...
+
+When using groupBy v2, three parameters control resource usage and limits:
+
+- druid.processing.buffer.sizeBytes: size of the off-heap hash table used for aggregation, per query, in bytes. At
+most druid.processing.numMergeBuffers of these will be created at once, which also serves as an upper limit on the
+number of concurrently running groupBy queries.
+- druid.query.groupBy.maxMergingDictionarySize: size of the on-heap dictionary used when grouping on strings, per query,
+in bytes. Note that this is based on a rough estimate of the dictionary size, not the actual size.
+- druid.query.groupBy.maxOnDiskStorage: amount of space on disk used for aggregation, per query, in bytes. By default,
+this is 0, which means all aggregation is done in-memory.
```
Comment (gianm) [inline docs/content/querying/groupbyquery.md]: It's both. Clarified.

File: processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java (hunk 1 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java]: This error will only be seen by the user if maxOnDiskStorage is 0, so "positive" does make sense. If it's > 0 they'll see a different error, about disk space being full.

File: processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java (hunk 2 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java]: Same as above

**Comment by gianm** [comment]
Re-pushed with some edits.

File: docs/content/querying/groupbyquery.md [hunk 5]
```diff
@@ -141,8 +141,11 @@ on-heap by default, but it can optionally store aggregated values off-heap.
 Query API and results are compatible between the two engines; however, there are some differences from a cluster
 configuration perspective:
 
-- groupBy v1 merges results in heap, whereas groupBy v2 merges results off-heap. As a result, optimal configuration for
-your Druid nodes may involve less heap (-Xmx, -Xms) and more direct memory (-XX:MaxDirectMemorySize).
+- groupBy v1 controls resource usage using a row-based limit (maxResults) whereas groupBy v2 uses bytes-based limits.
+In addition, groupBy v1 merges results on-heap, whereas groupBy v2 merges results off-heap. These factors mean that
+memory tuning and resource limits behave differently between v1 and v2. In particular, due to this, some queries
+that can complete successfully in one engine may exceed resource limits and fail with the other engine. See the
+"Memory tuning and resource limits" section for more details.
```
Comment (jon-wei) [inline docs/content/querying/groupbyquery.md]: nit: maybe turn this into a link to that section

**Pull request merged**